### ISSUE TITLE
Glad to see this repository, have some questions

### ISSUE THREAD

**Issue by rakybond007** [FreemanX/pocl-android-dependency#1]
Hello, I was trying to build pocl for android 28 (Galaxy S10), and glad to find this repo. If you are ok, I want to ask some questions.

1. In llvm build script for aarch64, you used Cmake. Is that Cmake from android-sdk, not a host's cmake right?.
2. And I want to know the llvm-tblgen path you used. What is /media/p~ directory means?
3. I want to know exact path in your README below,
build executable binaries for x86 and copy those files to llvm/bin/
build static libraries for aarch64 and put them in llvm/lib/
Is the llvm directory meaning ndk toolchain directory?

Thanks :)

**Comment by FreemanX** [comment]
Hi. I myself haven't successfully build pocl for aarch64 android yet either. This repo is for me to ask questions or to confirm that I'm on the right direction in pocl/pocl#807. Libraries built in this repo aren't finalized. Maybe we can share our progress and help each other on porting pocl to android.    

As for your question:  
1. llvm and clang must be the same as your android ndk compiler. You can get the compiler info via `aarch64-linux-android-clang --version`. Then you can clone and check out the exact same commit id as your ndk compiler. (I'm using ndk_r17 based on llvm6.0.2)           
2. llvm-tblgen path and clang-tblgen path are hardcoded to the host(your computer) build directory of LLVM. You need to build llvm and clang twice from the source code: (1) build for your host system. (2) Build for aarch64 with the same LLVM cmake settings. When you build for aarch64, it will automatically build tblgen. But that tblgen executable is for aarch64 and will be invoked later in the building process. Therefore, I manually direct llvm-tblgen and clang-tblgen to the host build directory.       
3. My compilation follows the logic of this script: https://github.com/pocl/pocl/blob/release_1_3/android/build-arm.sh. In this script, all dependencies are organized in one place, so did I. Note that llvm/bin are all host executables while llvm/lib are all aarch64 libraries. I'm not build static libraries for now, because later pocl compilation will fail when linking OpenCL.so. My android ndk compiler can't link that many static code into one file. This lead to an annoying problem that I have to manually export LD_LIBRARY_PATH every time.       


BTW:
```
binutils: pulled from android OS
adb pull /system/bin/ld.mc
mv ld.mc ld    
```
This won't work, I now build binutils from source myself.

**Comment by rakybond007** [comment]
Thank you for quick respond!! Your answer helped me much. I will try more tomorrow and If I have progress, I will share in here. Thanks :)

**Comment by rakybond007** [comment]
Hello, I tried to build llvm first today.
I first used ndk r21 and llvm which aarch64-linux-android-clang --version returned. My cmake version is 3.10.2
I succeed to build the llvm for host, and I passed the path of llvm-tblgen to build_llvm_aarch64.sh
(* I couldn't find clang-tblgen on install dir)
However, I got the error like

```
Built target llvm-tblgen
make[2]: *** No rule to make target `/home/hj/data2/llvm_x86_out/bin/llvm-tblgen', needed by `lib/IR/AttributesCompatFunc.inc.tmp'.  Stop.
make[1]: *** [lib/IR/CMakeFiles/AttributeCompatFuncTableGen.dir/all] Error 2
```
I thought that it is because of different version as you, So I downloaded android-ndk-r17c,
and I downloaded llvm and clang as aarch64-linux-android-clang --version returned.
llvm commit 34361f192e4 & clang commit 183abd29fc.

I succeed to build the llvm for host again, but I could not find clang-tblgen again and failed to build aarch64 with the same error...

Before, I only succeed to build the llvm for aarch64 with the reference of https://github.com/Jman420/llvm_for_android/blob/develop/build_llvm_android.sh
Followed format of above, my actual script was below.

`#!/bin/bash

AndroidSystemVersion="28"
AndroidSdkDir="/home/hj/ndk/android-sdk"
AndroidCmakeExe="$AndroidSdkDir/cmake/3.10.2.4988404/bin/cmake"
AndroidNinjaExe="$AndroidSdkDir/cmake/3.10.2.4988404/bin/ninja"
NdkBundle="$AndroidSdkDir/ndk/21.0.6113669"
ToolchainFile="$NdkBundle/build/cmake/android.toolchain.cmake"
ArchTargets=("arm64-v8a")
ArchTriples=("aarch64-linux-android")
LlvmTargets=("AArch64")

ProjectRootPath="android-project-llvm-default"
BuildRootPath="android-build-llvm-default"
SourceRootPath="llvm-project/llvm"

PatchableSourceFiles=( "$SourceRootPath/lib/Transforms/CMakeLists.txt" )
SourcePatchReplacePatterns=( "^add_subdirectory(Hello)$" )
SourcePatchReplaceValues=( "#add_subdirectory(Hello)" )

echo "Patching LLVM Source Files..."
PatchesLength=${#PatchableSourceFiles[*]}
for (( patchCounter=0; patchCounter < $PatchesLength; patchCounter++ ))
do
    patchFile=${PatchableSourceFiles[$patchCounter]}
    replacePattern=${SourcePatchReplacePatterns[$patchCounter]}
    replaceValue=${SourcePatchReplaceValues[$patchCounter]}
    
    echo "Patching Source File : $patchFile ..."
    if [ ! -f "$patchFile.orig" ]
    then
        cp "$patchFile" "$patchFile.orig"
    fi
    
    echo "Replacing $replacePattern with $replaceValue ..."
    sed -i "s/$replacePattern/$replaceValue/" "$patchFile"
    echo "Successfully patched Source File : $patchFile !"
done
echo "Successfully patched LLVM Source Files!"

echo "Building LLVM for Android..."
ArchsLength=${#ArchTargets[*]}
for (( archCounter=0; archCounter < $ArchsLength; archCounter++ ))
do
    archTarget=${ArchTargets[$archCounter]}
    archTriple=${ArchTriples[$archCounter]}
    llvmTarget=${LlvmTargets[$archCounter]}
    projectDir="$ProjectRootPath/$archTarget"
    buildDir="$BuildRootPath/$archTarget"
    
    echo "Setting Up Build Environment for Architecture : $archTarget ..."
    if [ -d $projectDir ]
    then
        echo "Removing existing Project directory : $projectDir ..."
        rm -rf "$projectDir"
    fi
    if [ -d $buildDir ]
    then
        echo "Removing existing Build directory : $buildDir ..."
        rm -rf "$buildDir"
    fi
    
    echo "Creating Project directory : $projectDir ..."
    mkdir -p $projectDir
    echo "Creating Build directory : $buildDir ..."
    mkdir -p $buildDir
    buildFullPath=$(realpath "./$buildDir")
    
    echo "Generating Project Files for Architecture : $archTarget ..."
    pushd $projectDir > /dev/null
    $AndroidCmakeExe \
        -DCMAKE_BUILD_TYPE="Release" \
        -DCMAKE_INSTALL_PREFIX="$buildFullPath" \
        -DCMAKE_TOOLCHAIN_FILE="$ToolchainFile" \
        -DCMAKE_MAKE_PROGRAM="$AndroidNinjaExe" \
        \
	-DANDROID_PLATFORM="$AndroidSystemVersion" \
	-DLLVM_ENABLE_PROJECTS="clang" \
        -DLLVM_TABLEGEN="/usr/bin/llvm-tblgen-8" \
        \
        -DLLVM_INCLUDE_BENCHMARKS=OFF \
        -DLLVM_INCLUDE_DOCS=OFF \
        -DLLVM_INCLUDE_EXAMPLES=OFF \
        -DLLVM_INCLUDE_RUNTIMES=OFF \
        -DLLVM_INCLUDE_TESTS=OFF \
        -DLLVM_INCLUDE_GO_TESTS=OFF \
        -DLLVM_INCLUDE_TOOLS=OFF \
        -DLLVM_INCLUDE_UTILS=OFF \
        \
        -DLLVM_BUILD_BENCHMARKS=OFF \
        -DLLVM_BUILD_EXAMPLES=OFF \
        -DLLVM_BUILD_RUNTIME=OFF \
        -DLLVM_BUILD_RUNTIMES=OFF \
        -DLLVM_BUILD_TESTS=OFF \
        -DLLVM_BUILD_TOOLS=OFF \
        -DLLVM_BUILD_UTILS=OFF \
        \
        -DLLVM_BUILD_LLVM_DYLIB=OFF \
        \
        -DLLVM_ENABLE_PIC=False \
        \
        -G "Ninja" \
        \
        "../../$SourceRootPath"
    if [ $? -ne 0 ]; then
        echo "Project Generation failed for Architecture : $archTarget !"
        popd > /dev/null
        exit 1
    fi
    
    echo "Building LLVM for Architecture : $archTarget ..."
    $AndroidCmakeExe --build . --target install
    if [ $? -ne 0 ]; then
        echo "Compilation failed for Architecture : $archTarget !"
        popd > /dev/null
        exit 1
    fi
    popd > /dev/null
    
    echo "Successfully built LLVM for Architecture : $archTarget !"
done
echo "Successfully built LLVM for Android!"
`
I want to know if you changed something to build or if something was different with you! Or did you tried to build in target device, not a cross compile on host? 
And in the script I used before, I could find it removing subdirectory "Hello". Did you do same? I succeed the script only in llvm-8, but It was only succeed case for me... Thanks :)

**Comment by FreemanX** [comment]
llvm-tblgen and clang-tblgen aren't located in the installing directory(install prefix), they are in your host build directory.      
Here's the my file hierarchy:
```
.
├── build        // host build dir
├── build-arm64  // aarch64 build dir
├── clang        // clang source code dir
└── llvm         // llvm source code dir
```
This line as well as the next
https://github.com/FreemanX/pocl-android-dependency/blob/6a3362864b70af61fe8472dac8ad6b9fc2094f1d/build_llvm_aarch64.sh#L41
both are pointing to build/bin/llvm-tblgen and build/bin/clang-tblgen (host build dir).     

I also tried the latest ndk before (even the latest from https://ci.android.com/builds/branches/aosp-ndk-release-r21/grid? ). The script should be able to compile llvm and clang successfully provided all variables are set correctly. The reason why I use r17 is merely for keeping the consistency with my other native projects and I know for sure that r17 works fine with my development environment setup(I'm still using Android 8 on my phone). If r21 works fine with you, you may keep using r21. Samsung S10 is newer than Oneplus 6 after all.  I didn't remove any "Hello" subdirectory.     
I didn't install llvm and clang to my host system either (apt install). tblgen must be the same version as your compiling ones.     `-DLLVM_TABLEGEN="/usr/bin/llvm-tblgen-8" \ `pointed out that the script is using the system installed tblgen which is version 8. That's why you can only success in llvm-8.

**Comment by rakybond007** [comment]
Oh, -DLLVM_TABLEGEN="/usr/bin/llvm-tblgen-8" \ this part was different every time as you said. When I tried to build llvm-9, I used llvm-tblgen-9. I have every versions in that directory. 

I thought I can find clang-tblgen in the path written in line 5 of build_llvm_x86.sh.
But as you said, I could find it in my build directory. Let me try again. Thanks everytime!

**Comment by rakybond007** [comment]
Hello, I could build other libraries with your help. Thanks :)
Now I have to build "ld" which is binutils. I cloned the link you mentioned, and checked the build.py in there.
However, I could not find any python modules on my ndk, Which is in (android-sdk/ndk), so I tried to find if there is other ndk for building "ld". Finally I found https://android.googlesource.com/platform/ndk/  and cloned it either. I could find some more python files on there different with my original ndk, but I couldn't find bootstrap.py yet.

`site.addsitedir(os.path.join(ANDROID_TOP, 'ndk'))
import bootstrap  # pylint: disable=import-error,wrong-import-position`

I guess you also used the build script, how did you import the module? It was not a prebuilt module, and also could not find on pip3 installable packages. 
I could find `import ndk.abis
from ndk.hosts import Host
import ndk.paths
import ndk.timer` this modules on ndk/ndk which was coned from above link. Thanks :)

**Comment by rakybond007** [comment]
Hello, Did you succeed to build the pocl now?
I could build every other libraries except pocl, binutil with your kind help :)
Now, I am configuring my build script of pocl based on your script, but I have Issues.

1. I remember that you had some linking problem and tried to build new ld. I think my problem is not exactly same with you, but I wonder if you had seen this kind of message either.
![image](https://user-images.githubusercontent.com/5442998/80923326-f19aa280-8dbd-11ea-998a-d43122d8611d.png)
Is it looks like your problem?
2. I also have other issue, 
![image](https://user-images.githubusercontent.com/5442998/80923344-1727ac00-8dbe-11ea-8b3b-25b12293c433.png)
I guess the iostream related everything shoud be on ndk sysroot, but this kind of error occurs... did you see this before?
Thanks Everytime...

**Comment by FreemanX** [comment]
1. You may check the CMakeLists.txt. POCL is asking for libdl and its header file dlfcn.h. I think they come with android ndk. You can find their path and give a hint to find_library and find_file functions.

2. I'm not sure what's happening in your case. It's something related to your toolchain path or compiling flags, I guess. You can create your own cpp file and try to compile and see what works. Notice that you should use the compiler come with the android ndk, not the one you compiled (from LLVM). I use static linking cpp library with flag `-static-libstdc++` instead of 
`-stdlib=libstdc++`.

**Comment by rakybond007** [comment]
Thanks :)
After I changed my build-pocl.sh linking clang of $ANDROID_TOOLCHAIN_SYSROOT_USR/bin/clang like
`ln -sf $ANDROID_TOOLCHAIN_BIN/$HOST\28-clang $ANDROID_TOOLCHAIN_SYSROOT_USR/bin/clang
ln -sf $ANDROID_TOOLCHAIN_BIN/$HOST\28-clang++ $ANDROID_TOOLCHAIN_SYSROOT_USR/bin/clang++`
This helped me to remove problem2.

I reported some Issue on pocl, and I found the reason was I didn't set `CMAKE_SYSTEM_PROCESSOR` which is needed in CMakeList.txt of PoCL!!!
I found this kind of setting is done in file which is indicated as `-DCMAKE_TOOLCHAIN_FILE=android.cmake`.
At first, I didn't give only that option different with you, because I didn't know what actual file I should use. But now I know I have to use it. So I tried to use [https://github.com/pocl/pocl/blob/release_1_3/android/androideabi.cmake](url) cmake file in here, and changed little bit from it, that changing compiler to clang and clang++.  However, it does not work, it has an error like `CMake Error at CMakeLists.txt:841 (pkg_check_modules):
  Unknown CMake command "pkg_check_modules".
`
Next, I tried to use android.toolchain.cmake file which I could find in `$NDK_PATH/build/cmake/android.toolchain.cmake`, (also set the variables like -DANDROID_ABI, or -DANDROID_PLATFORM in build-pocl.sh file) but when I used the file, it changed so many settings in build-pocl.sh Such as linking of compiler, setting environment variables... so that clang, clang++, llvm-as, llvm-link lead to incorrect file path.

So I wonder if you can share your android.cmake file which was used in pocl-build.sh... Thanks everytime

**Comment by FreemanX** [comment]
I also used https://github.com/pocl/pocl/blob/release_1_3/android/androideabi.cmake  with some changes made like you did to androideabi.cmake (change compiler, system processor settings...): 
```
SET(CMAKE_SYSTEM_NAME Linux)
SET(CMAKE_SYSTEM_VERSION 1)
SET(CMAKE_SYSTEM_PROCESSOR aarch64)

SET(CMAKE_C_COMPILER   aarch64-linux-android-clang)
SET(CMAKE_CXX_COMPILER aarch64-linux-android-clang++)

SET(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
SET(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
SET(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
```
